Sources/ComposableArchitecture/Beta/Concurrency.swift

```diff
@@ -0,0 +1,136 @@
+  ///
+  ///   var body: some View {
+  ///     WithViewStore(self.store) { viewStore in
+  ///       List {
+  ///         if let response = viewStore.response {
+  ///           Text(response)
+  ///         }
+  ///       }
+  ///       .refreshable {
+  ///         await viewStore.send(.pulledToRefresh, while: \.isLoading)
+  ///       }
+  ///     }
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// Here we've used the ``send(_:while:)`` method to suspend while the `isLoading` state is
+  /// `true`. Once that piece of state flips back to `false` the method will resume, signaling to
+  /// `.refreshable` that the work has finished which will cause the loading indicator to disappear.
+  ///
+  /// **Note:** ``ViewStore`` is not thread safe and you should only send actions to it from the
+  /// main thread. If you are wanting to send actions on background threads due to the fact that the
+  /// reducer is performing computationally expensive work, then a better way to handle this is to
+  /// wrap that work in an ``Effect`` that is performed on a background thread so that the result
+  /// can be fed back into the store.
+  ///
+  /// - Parameters:
+  ///   - action: An action.
+  ///   - predicate: A predicate on `State` that determines for how long this method should suspend.
+  public func send(
+    _ action: Action,
+    while predicate: @escaping (State) -> Bool
+  ) async {
+    self.send(action)
+    await self.suspend(while: predicate)
+  }
+
+  /// Sends an action into the store and then suspends while a piece of state is `true`.
+  ///
+  /// See the documentation of ``send(_:while:)`` for more information.
+  ///
+  /// - Parameters:
+  ///   - action: An action.
+  ///   - animation: The animation to perform when the action is sent.
+  ///   - predicate: A predicate on `State` that determines for how long this method should suspend.
+  public func send(
+    _ action: Action,
+    animation: Animation?,
+    while predicate: @escaping (State) -> Bool
+  ) async {
+    withAnimation(animation) { self.send(action) }
+    await self.suspend(while: predicate)
+  }
+
+  /// Suspends while a predicate on state is `true`.
+  ///
+  /// - Parameter predicate: A predicate on `State` that determines for how long this method should
+  /// suspend.
+  public func suspend(while predicate: @escaping (State) -> Bool) async {
+    var cancellable: Cancellable?
+    try? await withTaskCancellationHandler(
+      handler: { [cancellable] in cancellable?.cancel() },
+      operation: {
+        try Task.checkCancellation()
+        try await withUnsafeThrowingContinuation { (continuation: UnsafeContinuation<Void, Error>) in
+          guard !Task.isCancelled else {
+            continuation.resume(throwing: CancellationError())
+            return
+          }
+          cancellable = self.publisher
+            .filter { !predicate($0) }
+            .prefix(1)
+            .sink { _ in
+              continuation.resume()
+              _ = cancellable
```

Comment:
Out of curiosity, is this different from capturing `cancellable` with brackets like you do on l.116? Wouldn't
```swift
.sink { [cancellable] _ in continuation.resume() }
```
be equivalent?

Reply:
Good question. If we don't use the `cancellable` in the closure then we get an unused warning. So I think we gotta do the `_ = cancellable` dance.